File: ray-tracer/elements/things/Plane.cs
using ray_tracer.common;
using ray_tracer.scene;

namespace ray_tracer.elements.things
{
    public class Plane : Thing
    {
        public Surface surface { get; set; }

        public readonly Vector norm;
        public readonly double offset;

        public Plane(Vector norm, double offset, Surface surface)
        {
            this.norm = norm;
            this.offset = offset;
            this.surface = surface;
        }

        public Intersection intersect(Ray ray)
        {
            var normRayDirectionDot = norm.dot(ray.direction);

            if (normRayDirectionDot > double.Epsilon) return null;

            var dist = -(norm.dot(ray.origin) + offset) / normRayDirectionDot;

            return new Intersection(this, ray, norm, dist);
        }
    }
}

File: java-app/src/pl/mikolajczuk/kamil/raytracing/lib/elements/things/Plane.java
package pl.mikolajczuk.kamil.raytracing.lib.elements.things;

import pl.mikolajczuk.kamil.raytracing.lib.common.*;
import pl.mikolajczuk.kamil.raytracing.lib.scene.*;

import java.util.Optional;

public class Plane implements Thing {
    private final Surface surface;

    public Surface getSurface() {
        return this.surface;
    }

    public final Vector norm;
    public final double offset;

    public Plane(Vector norm, double offset, Surface surface) {
        this.norm = norm;
        this.offset = offset;
        this.surface = surface;
    }

    public Optional<Intersection> intersect(Ray ray) {
        var normRayDirectionDot = norm.dot(ray.direction);

        if (normRayDirectionDot > 0) return Optional.empty();

        var dist = -(norm.dot(ray.origin) + offset) / normRayDirectionDot;

        return Optional.of(new Intersection(this, ray, norm, dist));
    }
}

    